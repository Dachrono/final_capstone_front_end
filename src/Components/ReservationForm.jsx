import { useDispatch, useSelector } from 'react-redux';
import { Link, useParams } from 'react-router-dom';
import CurrentDate from '../logic/getDate';
import { reservation } from '../logic/formatObjects';
import { createReservation } from '../Redux/Slices/reservationsSlice';

const ReservationForm = () => {
  const dispatch = useDispatch();

  const { user } = useSelector((state) => state.users);
  const { experienceID } = useParams();
  const { experiences } = useSelector((state) => state.experiences);
  const selectedExperience = experiences.filter(
    (experience) => experience.id === parseInt(experienceID, 10),
  );

  const CreateReservation = (e) => {
    const btn = e.target;
    const dataForm = btn.parentElement;
    const city = dataForm.querySelector('#city');
    const date = dataForm.querySelector('#date');
    const msg = dataForm.querySelector('#message-error');

    if (date.value === '' || city.value === '') {
      msg.textContent = 'Fill out all the fields please';
    } else {
      const response = dispatch(createReservation(
        reservation(user.id, experienceID, date.value, city.value),
      ));

      response.then((data) => {
        if (typeof data.payload !== 'object') {
          msg.textContent = 'Something went wrong, please try again later';
        } else {
          msg.textContent = 'Reservation created successfully';
          city.value = '';
          date.value = '';
        }
      });
    }
  };

  return (
    <section className="image-back d-flex-col reserv-back">
      <div className="d-flex-col form-cont gap">
        <h2 className="title-form title-reserv color">
          Book
          {' '}
          {selectedExperience[0].name}
        </h2>
        <span className="d-flex reserv-desc span-desc">{selectedExperience[0].description}</span>
        <form className="d-flex-col gap" onSubmit={(ev) => ev.preventDefault()}>
          <span id="message-error" />
          <input id="city" type="text" placeholder="City" className="form-input reserv-input" required />
          <input id="date" type="datetime-local" min={CurrentDate()} max="2023-12-31 18:00" className="form-input reserv-input" required />
          <input type="submit" className="form-submit d-flex-col" onClick={(e) => CreateReservation(e)} value="Register" />
          <Link to="/MainPage" className="form-submit d-flex-col">Main page</Link>
        </form>
      </div>
    </section>
  );
};

export default ReservationForm;
