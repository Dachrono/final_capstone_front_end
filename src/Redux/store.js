import { configureStore } from '@reduxjs/toolkit';
import { logger } from 'redux-logger';
import usersReducer from './Slices/usersSlice';
import experiencesReducer from './Slices/ExperiencesSlice';
// import reservationsReducer from './ReservationsSlice/ReservationsSlice';

const store = configureStore({
  reducer: {
    users: usersReducer,
    experiences: experiencesReducer,
    // reservations: reservationsReducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(logger),
});

export default store;
