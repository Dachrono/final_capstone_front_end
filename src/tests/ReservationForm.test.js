import { Provider } from 'react-redux';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { render } from '@testing-library/react';
import store from '../assets/helpers/mockStore';
import ReservationForm from '../Components/ReservationForm';

jest.mock('../logic/getDate', () => ({
    __esModule: true,
    default: jest.fn(() => '2023-11-09T23:03:00.000Z'),
}));

test('RESERVATION FORM', () => {
  const tree = render(
    <Provider store={store}>
      <MemoryRouter initialEntries={['/Experience1/1/new-reservation']}>
        <Routes>
          <Route path="/:experienceName/:experienceID/new-reservation" element={<ReservationForm />} />
        </Routes>
      </MemoryRouter>
    </Provider>,
  );

  expect(tree).toMatchSnapshot();
});