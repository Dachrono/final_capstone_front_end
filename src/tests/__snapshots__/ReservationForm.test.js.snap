// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RESERVATION FORM 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="image-back d-flex-col reserv-back"
      >
        <div
          class="d-flex-col form-cont gap"
        >
          <h2
            class="title-form title-reserv color"
          >
            Book
             
          </h2>
          <span
            class="d-flex reserv-desc span-desc"
          />
          <form
            class="d-flex-col gap"
          >
            <span
              id="message-error"
            />
            <input
              class="form-input reserv-input"
              id="city"
              placeholder="City"
              required=""
              type="text"
            />
            <input
              class="form-input reserv-input"
              id="date"
              max="2023-12-31 18:00"
              required=""
              type="datetime-local"
            />
            <input
              class="form-submit d-flex-col"
              type="submit"
              value="Register"
            />
          </form>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="image-back d-flex-col reserv-back"
    >
      <div
        class="d-flex-col form-cont gap"
      >
        <h2
          class="title-form title-reserv color"
        >
          Book
           
        </h2>
        <span
          class="d-flex reserv-desc span-desc"
        />
        <form
          class="d-flex-col gap"
        >
          <span
            id="message-error"
          />
          <input
            class="form-input reserv-input"
            id="city"
            placeholder="City"
            required=""
            type="text"
          />
          <input
            class="form-input reserv-input"
            id="date"
            max="2023-12-31 18:00"
            required=""
            type="datetime-local"
          />
          <input
            class="form-submit d-flex-col"
            type="submit"
            value="Register"
          />
        </form>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
